rancherdesktop/instanceInfo.go

```diff
@@ -0,0 +1,81 @@
+package rancherdesktop
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+)
+
+var validPlatforms map[string]struct{} = map[string]struct{}{
+	"win32":  {},
+	"darwin": {},
+	"linux":  {},
+}
+
+var validArchs map[string]struct{} = map[string]struct{}{
+	"x64":   {},
+	"arm64": {},
+}
+
+type CheckUpgradeRequest struct {
+	AppVersion string            `json:"appVersion"`
+	ExtraInfo  map[string]string `json:"extraInfo"`
+}
+
+// InstanceInfo contains all the info we need about an instance of Rancher Desktop.
+// Note that older versions of Rancher Desktop cannot be represented by
+// an InstanceInfo struct, since they do not send certain required fields.
+type InstanceInfo struct {
+	AppVersion      *semver.Version
+	Platform        string
+	Arch            string
+	PlatformVersion *semver.Version
+}
+
+// NewInstanceInfo converts the general CheckUpgradeRequest type into an InstanceInfo.
+// If the CheckUpgradeRequest does not contain the needed info (which is optional in
+// a CheckUpgradeRequest), an error is returned.
+func NewInstanceInfo(checkUpgradeRequest *CheckUpgradeRequest) (InstanceInfo, error) {
+	appVersion, err := semver.NewVersion(checkUpgradeRequest.AppVersion)
```

Comment:
It is passed as a pointer because that's how the original code did it, and I didn't notice that it could be improved.

I agree though. While it can't be nil (it is initialized as a value in `*Server.CheckUpgrade()` and then passed as a pointer to `*Server.CheckUpgradeResponse`), there is no need for passing it as a pointer.

> As a general rule I would say all parameters should be passed by value, unless you either need to modify it in-place, or you need an out-of-band nil value.

I agree, except for the case where one or more of the methods of a type take a pointer to that type in the receiver argument. In those cases it seems (I could be wrong) that it is typical to just work with a pointer to that type.